staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -693,19 +698,43 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
 	return false
 }
 
-func ensureNodeProvidedIPExists(node *v1.Node, nodeAddresses []v1.NodeAddress) (*v1.NodeAddress, bool) {
-	var nodeIP *v1.NodeAddress
-	nodeIPExists := false
-	if providedIP, ok := node.ObjectMeta.Annotations[cloudproviderapi.AnnotationAlphaProvidedIPAddr]; ok {
-		nodeIPExists = true
-		for i := range nodeAddresses {
-			if nodeAddresses[i].Address == providedIP {
-				nodeIP = &nodeAddresses[i]
-				break
+// preferProvidedNodeIP re-orders node addresses, placing a nodeIP suggested by
+// the user first. If the user suggested a node IP which is not present in node
+// addresses it will indicate an error by also returning false.
+func preferProvidedNodeIP(node *v1.Node, nodeAddresses []v1.NodeAddress) ([]v1.NodeAddress, error) {
+	providedIP, ok := node.ObjectMeta.Annotations[cloudproviderapi.AnnotationAlphaProvidedIPAddr]
+	if !ok {
+		// No suggested node IP
+		return nodeAddresses, nil
+	}
+	klog.V(2).Infof("Kubelet provided node IP %s. Ensuring it is listed as first node address.", providedIP)
+
+	findIndex := func(address string) (int, bool) {
+		for i, nodeAddress := range nodeAddresses {
+			if nodeAddress.Address == address {
+				return i, true
 			}
 		}
+		return 0, false
 	}
-	return nodeIP, nodeIPExists
+
+	index, nodeIPFound := findIndex(providedIP)
+	if !nodeIPFound {
+		// Node IP was suggested but is not valid
+		return nil, fmt.Errorf("node IP %s was provided by kubelet for node %s, but it is not present in the cloud provider", providedIP, node.Name)
+	}
+
+	if index == 0 {
+		// Suggested node IP is already first
+		return nodeAddresses, nil
+	}
+
+	// Move the suggested address to the front by moving those before it right one position
+	primary := nodeAddresses[index]
+	copy(nodeAddresses[1:index+1], nodeAddresses[0:index])
```

Comment:
Had a conversion with Jordan today about the implications of this change. Overall it seems correct that node controller sets the order of the address, but it should set the order only scoped to the address type it belongs to. E.g. if --node-ip matches against an InternalIP, it should sort it such that it is the first InternalIP in status but not necessary the first amongst all adddresses (unless it was already the case).

Reply:
Thanks. I've just pushed a new commit (well 2, but one of them is just readability).

Incidentally, it's only a couple of lines of code but I personally feel future me would appreciate the verbose comment, especially if I hadn't had coffee yet. Hope it doesn't fall foul of style guidelines.

Reply:
Wouldn't it make more sense to extract the kubelet cloud-provider-node-ip-addresses-reordering code to `"k8s.io/component-helpers/node/util"` and then use that code here?

Reply:
I don't imagine there would be any other users of it, or are you thinking it would just be tidier?

Reply:
There wouldn't be any other users. And I guess kubelet's usage of it will eventually go away when the legacy providers go away... So maybe just copy the code from there to here instead?

It's just, if we want the same behavior, let's have the same code. (The initial version of the code here didn't do the right thing... are we sure the current version does?)

Reply:
Now I understand: have kubelet legacy and cloud-provider use the same re-ordering code, which is in node/util. I'll look into it.